app/src/org/commcare/android/database/app/models/FormDefRecordV12.java

```diff
@@ -0,0 +1,155 @@
+package org.commcare.android.database.app.models;
+
+import android.database.Cursor;
+import android.database.SQLException;
+
+import org.apache.commons.lang3.StringUtils;
+import org.commcare.android.storage.framework.Persisted;
+import org.commcare.models.database.SqlStorage;
+import org.commcare.models.framework.Persisting;
+import org.commcare.modern.database.Table;
+import org.commcare.modern.models.MetaField;
+import org.commcare.provider.FormsProviderAPI;
+import org.commcare.util.LogTypes;
+import org.commcare.utils.FileUtil;
+import org.javarosa.core.services.Logger;
+
+import java.io.File;
+import java.io.IOException;
+
+import static org.commcare.android.database.app.models.FormDefRecord.META_DISPLAY_NAME;
+import static org.commcare.android.database.app.models.FormDefRecord.META_FORM_FILE_PATH;
+import static org.commcare.android.database.app.models.FormDefRecord.META_FORM_MEDIA_PATH;
+import static org.commcare.android.database.app.models.FormDefRecord.META_JR_FORM_ID;
+import static org.commcare.android.database.app.models.FormDefRecord.META_MODEL_VERSION;
+import static org.commcare.android.database.app.models.FormDefRecord.META_UI_VERSION;
+
+/**
+ * Represents the version of a FormDefRecord that exists on any devices running a pre-2.47
+ * version of CommCare, which was deprecated in app db version 13. This class is used to read a
+ * FormDefRecord that exists in such a database, in order to run a db upgrade.
+ *
+ */
+@Table(FormDefRecord.STORAGE_KEY)
+public class FormDefRecordV12  extends Persisted {
+
+    @Persisting(1)
+    @MetaField(META_DISPLAY_NAME)
```

Comment:
Quick note that it's generally better to inline these as raw strings (IE: META_DISPLAY_NAME -> "displayName") so that a change to the static definition won't affect the migration path.

Reply:
Good point 👍

Reply:
nice, good idea!